### PR TITLE
least (copy), greatest (copy) and dateadd mock

### PR DESCRIPTION
**Pull request by osipovartem**
- Copied **least** and **greatest** UDF functions from this [PR](https://github.com/apache/datafusion/pull/13786) (already merged, but not released). After datafusion release we should remove this local copy
- Added mock for **dateadd** function to support sql syntax
- Changed logic for datafusion -> snowflake timestamps convertor for a while (still receive the error _252005: Failed to convert: field min_last_success: TIMESTAMP_NTZ::min_last_success, Error: invalid literal for int() with base 10: 'min_last_success'_)

### CODE REVIEW

**Comment by DanCodedThis** [comment]
For dateadd we can either use utf8 or it's variants or extend the sql parser to actualy work with underlying types, but the addition operator should work out of the box, more info in issue #68 (keep in mind there are two branhces now)

**Comment by DanCodedThis** [comment]
Also when using nanoseconds, maybe I'm using to small of a value but I can't seem to see the change, or is it a part of that "thing" where timestamps don't work with them yet?

**Comment by DanCodedThis** [comment]
And us (microsenconds) also

File: crates/control_plane/src/sql/functions/common.rs [hunk 1]
```diff
@@ -81,13 +77,14 @@ pub fn convert_record_batches(
...
 fn convert_timestamp_to_struct(column: &ArrayRef, unit: &TimeUnit) -> ArrayRef {
+    let now = Utc::now().timestamp();
     let (epoch, fraction) = match unit {
         TimeUnit::Second => {
             let array = column
                 .as_any()
                 .downcast_ref::<TimestampSecondArray>()
                 .unwrap();
-            let epoch: Int64Array = array.clone().unary(|x| x);
+            let epoch: Int64Array = array.iter().map(|x| x.unwrap_or(now)).collect();
```
Comment (eadgbear) [inline crates/control_plane/src/sql/functions/common.rs]: You might be able to use `map_or` here

**Review by eadgbear — DISMISSED** [review]

File: crates/control_plane/src/sql/functions/common.rs (hunk 1 above)
Comment (DanCodedThis) [reply crates/control_plane/src/sql/functions/common.rs]: I'll look into it on Monday or Artem will

**Comment by rampage644** [comment]
Closes #75

**Comment by DanCodedThis** [comment]
i64 max is 9_223_372_036_854_775_807, and since we are adding nanoseconds even with years, if somebody decides to add 30 years to any timestamp/date/time we will have an overflow (in the value that is being added, not the timestamp), we are flying dangerously close to the sun for my liking, or are we?

I initially decided not to use MonthDayNano since days and month are i32, and even though we distinguish years, month, weeks, etc., anybody by mistake can write any value that may be bigger than i32 max, accidentally (since we always receive i64, ex.: was writing nanoseconds switched the form ns to yy) or is it just an edge case?

Plus the other problem is leap years, and "leap months", which should be handled internally by timestamps (didn't check yet, but it should be common sense), but because we are adding nanoseconds, our months are always 30 days and our years are always 365 days. 

I'm going to look into this, and after that the PR should be ready

**Review by DanCodedThis — APPROVED** [review]
Dateadd completed

**Review by DanCodedThis — DISMISSED** [review]
Dateadd completed

File: crates/control_plane/src/sql/functions/common.rs [hunk 2]
```diff
@@ -81,60 +82,116 @@ pub fn convert_record_batches(
...
-        }
+    let now = Utc::now();
+    let timestamps: Vec<_> = match unit {
+        TimeUnit::Second => column
+            .as_any()
+            .downcast_ref::<TimestampSecondArray>()
+            .unwrap()
+            .iter()
+            .map(|x| {
+                let ts = DateTime::from_timestamp(x.unwrap_or(now.timestamp()), 0).unwrap();
```
Comment (eadgbear) [inline crates/control_plane/src/sql/functions/common.rs:93]: Is now a reasonable default if `x` can't be unwrapped?

**Review by eadgbear — APPROVED** [review]

File: crates/control_plane/src/sql/functions/common.rs (hunk 2 above)
Comment (DanCodedThis) [reply crates/control_plane/src/sql/functions/common.rs:93]: Maybe zero the unix 1970 timestamp?

Comment (DanCodedThis) [reply crates/control_plane/src/sql/functions/common.rs:93]: For now let it be. Can't merge the base branch is protected

**Pull request merged**